### ISSUE TITLE
Adding labels instead of circle markers for place

### ISSUE THREAD

**Issue by albanm** [Leaflet/Leaflet.VectorGrid#67]
Hello,

I can't find this in the documentation. Is it possible to add points with some textual property as leaflet labels (or small texts of anykind) instead of markers ?

I guess it should be possible to use the style function to add an item in a label layer instead of returning a style. But I don't see a way to access the coordinates of the point in this function.

Thanks a lot for this awesome project !

**Comment by perliedman** [comment]
Hi @albanm,
this is not currently possible - points can be styled as `CircleMarker`s or `Icon`s, nothing else.

Doing what you propose would be somewhat tricky: VectorGrid uses tiles with SVG or Canvas rendering, and naively adding labels to those would most certainly clip them in an undesirable way. Adding them as normal Leaflet layers would, on the other hand, change much of how VectorGrid basic design (a tile _grid_).

What I think we _could_ do, is exposing something similar to `L.GeoJSON`'s `oneachfeature` option, to allow users to hook on any logic they want; for that, we would also need a mechanism so that users can also clean up when features in tiles are removed.

**Comment by IvanSanchez** [comment]
@perliedman If a point can be styled with an `Icon`, can a `DivIcon` be used instead? If so, there's the text labels.

**Comment by perliedman** [comment]
@IvanSanchez no, at the moment we just take the icon image and draw it onto the canvas (or use an SVG `image`), so we're not even handling shadows or so...

I perceive VectorGrid's primary use case as "render a lot of things with decent performance", and adding one or more DOM elements per feature seems to go against this. But I realize there might be other reasons for using VectorGrid as well.

**Comment by albanm** [comment]
Vector tiles are sure great for rendering lots of features. But what I am interested in mostly is client side styling and intelligent simplification. Anyway even with lots of features not all layers are equally dense, for example the 'place' or 'poi' layers in osm tiles are ok.

Providing a hook to let the client add whatever it wants on top of the features in the tiles would be great !

I am willing to work on a PR for this.

**Comment by JulieGoldberg** [comment]
Was a workaround for this ever implemented? I'd like to use it if so.

**Comment by ABPW10** [comment]
Use SVG's `<text>` tag to write your labels, then pass the encoded SVG data URL string as the point's `icon`'s `iconUrl` option, then you will be able to see labels for the points.

Example Code:
```
function style(properties){
          const fs=10+properties.level
          const l=properties.name.length
          const dy=11.5
          const h=fs+dy,w=fs*l
          return {icon:L.icon({iconUrl:"data:image/svg+xml;base64," + btoa(unescape(encodeURIComponent(`<svg xmlns="http://www.w3.org/2000/svg" height="${h}" width="${w}">
<circle cx="${w/2}" cy="${h-7}" r="2" stroke="black" stroke-width="1" fill="white" />
<text paint-order="stroke fill" fill="white" stroke="black" stroke-width="2" x="50%" y="100%" font-size="${fs}px" text-anchor="middle" alignment-baseline="baseline" dy="-${dy}">${n}</text>
</svg>`))),iconSize:[w,h],iconAnchor:[w/2,h-7]})}
}
```
Remember to adjust the expression for caculating label width, `fs*l` is based on full-width 2-byte characters ( Chinese ) that I work with, for Latin characters in monospaced font it is `fs*l/2`. For non-monospaced fonts maybe canvas's `measureText` method can be used to figure out the dimensions of texts?

Then add a CSS to make labels that are out of the tile's bounds visible and add a `className` option to your vector tile layer.
```
    const vtOutOfBoundStyle=document.createElement('style')
    vtOutOfBoundStyle.textContent='.vt-labels svg{overflow:visible}'
    document.body.appendChild(vtOutOfBoundStyle)
```

Labels create in this way are not decluttered, meaning labels close to each other will be displayed one over another, but you can customize L.LayerGroup.Collision or similar plugins to make it work with labels for vector tiles created in the above way.
